src/realm/object-store/sync/async_open_task.cpp

```diff
@@ -118,4 +117,68 @@ void AsyncOpenTask::unregister_download_progress_notifier(uint64_t token)
         m_session->unregister_progress_notifier(token);
 }
 
+void AsyncOpenTask::run_subscription_initializer(SubscriptionCallback&& subscription_callback,
+                                                 AsyncOpenCallback&& async_open_callback,
+                                                 std::shared_ptr<_impl::RealmCoordinator> coordinator,
+                                                 bool rerun_subscription_on_open)
+{
+    auto shared_realm = coordinator->get_realm();
+    auto latest_subscription = shared_realm->get_latest_subscription_set();
+    static std::atomic<bool> subscription_state = false;
+
+    // we can end up here into 2 cases:
+    //  1. this is the first time we are opening/creating a realm, in this case the subscription version must be 0
+    //  2. the realm file has already been created, but the SDK has instructed us to run the subscription initializer.
+    //
+    //  In case 2. we don't want to run the subscription callback multiple times, so we keep some sort of state in
+    //  order to minimize this possibility. However there is still a chance that the subscription could fail and the
+    //  callback not run, as long as we don't receive acknowledgment of this from the server.
+    if (latest_subscription.version() == 0 || (rerun_subscription_on_open && !subscription_state.load())) {
```

Comment:
@danieltabacaru I added the state for detecting if the subscription was invoked already for realms that were already opened. The assumption is that the subscription will succeed, if it does not, then we reset the state, and we let the caller run the subscription the next time async open will be called.

Reply:
this does not longer apply.